### PR TITLE
feat: tile filters

### PR DESCRIPTION
**Pull request by IrakliJani**
closes: https://github.com/lightdash/lightdash/issues/3657
closes: https://github.com/lightdash/lightdash/issues/2346
closes: https://github.com/lightdash/lightdash/issues/1617

### Description:

⚠️ before testing in staging: open console, run `localStorage.setItem('dashboard_filters', true)` and refresh the browser.

- [x] You should be able to select/deselect the tiles that you would like to apply your chosen filter to. You should only be able to choose from tiles that include the relevant field type. Eg. if I choose a date filter but Tile E is made from a model that doesn’t have date fields, it should not appear as a selectable option.
- [x] You should be able to choose which column you want to filter by per tile. You should only be able to choose from the same column type. eg dates!
- [x] You should be able to access the ‘tiles’ tab in view mode as well as edit mode.

video demo:

https://user-images.githubusercontent.com/962095/203396228-7ff85ecd-121d-40de-ac46-3ff33a746fb0.mp4

### CODE REVIEW

**Comment by IrakliJani** [comment]
- [ ] need to response validation
CleanShot 2022-11-22 at 20 42 57@2x

File: packages/frontend/src/hooks/dashboard/useDashboardFiltersForExplore.ts [hunk 1]
```diff
@@ -1,26 +1,62 @@
...
+                    const tileConfig = f.tileConfigs?.find(
+                        (t) => t.tileUuid === tileUuid,
+                    );
+                    if (!tileConfig) return f;
+
+                    return {
+                        ...f,
+                        target: {
+                            fieldId: tileConfig.fieldId,
+                            tableName: tileConfig.fieldId.split('_')[0], // TODO: use a better way to get a table name
```
Comment (IrakliJani) [inline packages/frontend/src/hooks/dashboard/useDashboardFiltersForExplore.ts:40]: need to fix this.

**Comment by IrakliJani** [comment]
- [ ] add e2e tests

**Comment by IrakliJani** [comment]
- [ ] duplicated field edge case with underlying data


https://user-images.githubusercontent.com/962095/203569969-896297f7-2988-4d8e-aee9-df12bde1e359.mp4

**Comment by Martin-Carlsson** [comment]
Really cool!!

**Comment by TuringLovesDeathMetal** [comment]
@IrakliJani I'm not sure that this microcopy is true 🤔 

image

I _**think**_ that filters are added to the filters on the underlying charts, right?

**Comment by TuringLovesDeathMetal** [comment]
I'm trying to test the filters in the demo environment link here, but I don't think they exist in there yet 😅 

Where is the most helpful place for me to look to try to test things out @IrakliJani ?

**Comment by IrakliJani** [comment]
@TuringLovesDeathMetal oops, it's behind the "feature flag". open console in the browser and run `localStorage.setItem('dashboard_filters', true)` and then refresh the page

File: packages/frontend/src/components/DashboardFilter/FilterConfiguration/index.tsx [hunk 2]
```diff
@@ -1,99 +1,206 @@
...
 
     return (
         <ConfigureFilterWrapper>
-            {onBack && (
-                <BackButton minimal onClick={onBack}>
-                    Back
-                </BackButton>
+            <FieldLabel item={field} />
+
+            {localStorage.getItem('dashboard_filters') ? (
```
Comment (owlas) [inline packages/frontend/src/components/DashboardFilter/FilterConfiguration/index.tsx:150]: will we remove this before release?

File: packages/common/src/types/filter.ts [hunk 3]
```diff
@@ -66,12 +66,18 @@ export type DashboardFieldTarget = {
...
+};
+
 export type DashboardFilterRule<
     O = FilterOperator,
     T = DashboardFieldTarget,
     V = any,
     S = any,
-> = FilterRule<O, T, V, S>;
+> = FilterRule<O, T, V, S> & {
+    tileTargetOverride?: TileFieldTargetOverride[];
```
Comment (owlas) [inline packages/common/src/types/filter.ts:79]: could this be a map `Record<string,DashboardFieldTarget>` since it doesn't make sense to have the same tile uuid twice? It would also remove the additional type

**Comment by TuringLovesDeathMetal** [comment]
Niiiice!

I've played around with it a bit and it looks like everything is working great :) 

The only tiny thing is the copy when you're adding filters (like I mentioned above)

File: packages/frontend/src/hooks/dashboard/useDashboardFiltersForExplore.ts [hunk 4]
```diff
@@ -1,26 +1,63 @@
...
 import { useDashboardContext } from '../../providers/DashboardProvider';
 
 const useDashboardFiltersForExplore = (
+    tileUuid: string,
     explore: Explore | undefined,
 ): DashboardFilters => {
     const { dashboardFilters, dashboardTemporaryFilters } =
         useDashboardContext();
 
+    const tables = useMemo(
```
Comment (owlas) [inline packages/frontend/src/hooks/dashboard/useDashboardFiltersForExplore.ts:16]: in my next PR I'll move this to the backend

File: packages/frontend/src/hooks/dashboard/useDashboardFiltersForExplore.ts [hunk 5]
```diff
@@ -1,26 +1,63 @@
...
+        () => (explore ? Object.keys(explore.tables) : []),
+        [explore],
+    );
+
+    const overrideTileFilters = useCallback(
```
Comment (owlas) [inline packages/frontend/src/hooks/dashboard/useDashboardFiltersForExplore.ts:21]: I think this logic would also be simpler if the `tileTargetOverride` was a map.

It feels like the `DashboardFilterRule` has a dual purpose, it holds the configuration for the dashboard filters but then also the filters themselves.

i.e. the moment where we go:

```
...f,
target: {
  fieldId: tileConfig.fieldId,
  tableName: tileConfig.tableName
}
```

This still has the `tileTargetOverride` property, which looks a bit strange.


I feel like this function should take the configuration (includes tile overrides) and spit out a set of valid filters (not knowing anything about tile overrides)

**Review by owlas — CHANGES_REQUESTED** [review]
Since there isn't referential integrity between the tile uuids and the filter overrides, we could update the model to keep this as clean as possible.

Anytime a version is created in `DashboardModel.createVersion()` then we can prune the filter overrides and remove any irrelevant ones.

These lines:

```typescript
    private static async createVersion(
        trx: Transaction,
        dashboardId: number,
        version: DashboardVersionedFields,
    ): Promise<void> {
        const [versionId] = await trx(DashboardVersionsTableName).insert(
            {
                dashboard_id: dashboardId,
                updated_by_user_uuid: version.updatedByUser?.userUuid,
            },
            ['dashboard_version_id', 'updated_by_user_uuid'],
        );
        await trx(DashboardViewsTableName).insert({
            dashboard_version_id: versionId.dashboard_version_id,
            name: 'Default',
            filters: version.filters || {
                dimensions: [],
                metrics: [],
            },
        });
```

File: packages/frontend/src/components/DashboardFilter/FilterConfiguration/index.tsx (hunk 2 above)
Comment (IrakliJani) [reply packages/frontend/src/components/DashboardFilter/FilterConfiguration/index.tsx:150]: I don't think so, because we still need to fix few issues with UI + there are some follow up tasks.

File: packages/common/.eslintrc.js [hunk 6]
```diff
@@ -20,5 +20,6 @@ module.exports = {
             'warn',
             { argsIgnorePattern: '^_' },
         ],
+        'import/no-cycle': ['warn', { maxDepth: 1 }],
```
[the remark this answers is not in the document]
Comment (IrakliJani) [reply packages/common/.eslintrc.js]: not relevant anymore. this was happening in common when importing types between two files

File: packages/common/src/types/filter.ts (hunk 3 above)
Comment (IrakliJani) [reply packages/common/src/types/filter.ts:79]: updated

File: packages/common/src/types/filter.ts [hunk 7]
```diff
@@ -66,12 +69,18 @@ export type DashboardFieldTarget = {
     tableName: string;
 };
 
+export type TileFieldTargetOverride = DashboardFieldTarget & {
+    tileUuid: string;
+};
+
 export type DashboardFilterRule<
```
[the remark this answers is not in the document]
Comment (IrakliJani) [reply packages/common/src/types/filter.ts]: @owlas it's still relevant, because not all tiles override the default target. we can remove the `target` and `tileTargetOverride` if we replace it with `targets: Record<string, DashboardFieldTarget>` but then we need to have a migration for backward compatibility or do the ad-hoc magic in the frontend. thoughts?

**Comment by IrakliJani** [comment]
> @IrakliJani I'm not sure that this microcopy is true 🤔
> 
> image
> 
> I _**think**_ that filters are added to the filters on the underlying charts, right?

@TuringLovesDeathMetal right.. dashboard filters are added to the chart filters.
any suggestions for the copy?

**Comment by TuringLovesDeathMetal** [comment]
What about:

`These filters will be added to existing chart filters.`

Or, do we even need copy here? 🤔 there must be a reason why we added it, but I'm not sure

**Comment by IrakliJani** [comment]
@TuringLovesDeathMetal maybe in one sentence something like "Dashboard filters are added to and wont override existing chart filters" 🤔

**Comment by TuringLovesDeathMetal** [comment]
Ya...I wonder if we can just remove it? 

I'm going to post in Slack to see what peeps think 😅

**Comment by PriPatel** [comment]
@TuringLovesDeathMetal 
This can be removed! I remember adding this text - I don't think its adding any value (and it's wrong as well it turns out😂).
Lets clean up the clutter!

**Comment by TuringLovesDeathMetal** [comment]
https://media.tenor.com/KJgAfNXWdFEAAAAC/thumbs-down.gif

**Comment by TuringLovesDeathMetal** [comment]
image

**Comment by TuringLovesDeathMetal** [comment]
I just spent like 5 mins trying to figure out how to embed a gif into a comment...but I can't 😓 

So here's my lame gif-link: https://media.tenor.com/KJgAfNXWdFEAAAAC/thumbs-down.gif

File: packages/common/src/index.ts [hunk 8]
```diff
@@ -393,8 +395,64 @@ export const createFilterRuleFromField = (
...
+    a.type === b.type;
+
+const getDefaultTileFieldTargetOverride = (
+    field: FilterableField,
+    tilesSavedQueryFilters: Record<string, FilterableField[]>,
+): TileFieldTargetOverride[] =>
+    Object.entries(tilesSavedQueryFilters)
+        .map<TileFieldTargetOverride | undefined>(
+            ([tileUuid, savedQueryFilters]) => {
+                const filterableField = savedQueryFilters.find(
```
[the remark this answers is not in the document]
Comment (IrakliJani) [reply packages/common/src/index.ts]: this is my bad, actual name of the variable should be `availableFilters` not `savedQueryFilters` (renamed). 

that function in particular finds tiles(saved queries) which have the exact match with dashboard filter enables filters for them by default

File: packages/common/src/index.ts [hunk 9]
```diff
@@ -393,8 +395,64 @@ export const createFilterRuleFromField = (
         value ? [value] : [],
     );
 
+export const matchFieldExact = (a: FilterableField) => (b: FilterableField) =>
```
[the remark this answers is not in the document]
Comment (IrakliJani) [reply packages/common/src/index.ts]: extracted everything related to filters from index to `utils/filters.ts`

**Comment by IrakliJani** [comment]
fixing the backend test now

File: packages/frontend/src/hooks/dashboard/useDashboardFiltersForExplore.ts (hunk 5 above)
Comment (owlas) [reply packages/frontend/src/hooks/dashboard/useDashboardFiltersForExplore.ts:21]: What is the type returned by this callback?

**Review by owlas — APPROVED** [review]

**Comment by IrakliJani** [comment]
only the headless browser e2e is failing. merging...
